File: tests/PostProcessingTest.cxx
/*!
 * \file   tests/PostProcessingTest.cxx
 * \brief
 * \author Thomas Helfer
 * \date   04/02/2022
 * \copyright (C) Copyright Thomas Helfer 2018.
 * Use, modification and distribution are subject
 * to one of the following licences:
 * - GNU Lesser General Public License (LGPL), Version 3.0. (See accompanying
 *   file LGPL-3.0.txt)
 * - CECILL-C,  Version 1.0 (See accompanying files
 *   CeCILL-C_V1-en.txt and CeCILL-C_V1-fr.txt).
 */

#include <cmath>
#include <array>
#include <cstdlib>
#include <iostream>
#include <string_view>
#include "MGIS/Behaviour/Behaviour.hxx"
#include "MGIS/Behaviour/BehaviourData.hxx"
#include "MGIS/Behaviour/MaterialDataManager.hxx"
#include "MGIS/Behaviour/Integrate.hxx"

bool success = true;

bool check(const bool b, const std::string_view msg) {
  if (!b) {
    success = false;
    std::cerr << msg << '\n';
  }
  return b;
}

void check_behaviour(const mgis::behaviour::Behaviour& b,
                     const mgis::behaviour::Hypothesis h) {
  using namespace mgis::behaviour;
  check(b.behaviour == "PostProcessingTest", "invalid behaviour name");
  check(b.hypothesis == h, "invalid hypothesis");
  check(b.source == "PostProcessingTest.mfront", "invalid source");
  check(b.tfel_version == TFEL_VERSION, "invalid TFEL version");
  check(b.postprocessings.size() == 1u, "invalid number of post-processings");
  check(b.postprocessings.find("PrincipalStrain") != b.postprocessings.end(),
        "invalid post-processing name");
  const auto& p = b.postprocessings.at("PrincipalStrain");
  check(p.outputs.size() == 1u, "invalid post-processing output size");
  const auto& o = p.outputs.at(0);
  check(o.name == "PrincipalStrain", "invalid post-processing output");
  check(o.type == Variable::VECTOR_3D, "invalid post-processing output type");
  check(o.type_identifier == 2 + (3 << 3),
        "invalid post-processing output type identifier");
  check(getVariableSize(o, h) == 3, "invalid post-processing output size");
}  // end of check_behaviour

void call_postprocessing(const mgis::behaviour::Behaviour& b) {
  using namespace mgis::behaviour;
  constexpr auto e =
      std::array<mgis::real, 6u>{1.3e-2, 1.2e-2, 1.4e-2, 0., 0., 0.};
  constexpr auto e2 =
      std::array<mgis::real, 6u>{1.2e-2, 1.3e-2, 1.4e-2, 0., 0., 0.};
  constexpr auto eps = 10 * std::numeric_limits<mgis::real>::epsilon();
  auto d = BehaviourData{b};
  // initialize the states
  setExternalStateVariable(d.s0, "Temperature", 293.15);
  setExternalStateVariable(d.s1, "Temperature", 293.15);
  //
  for (mgis::size_type i = 0; i != 6; ++i) {
    d.s1.gradients[i] = e[i];
  }
  //
  auto outputs = allocatePostProcessingVariables(b, "PrincipalStrain");
  if (!check(outputs.size() == 3u, "invalid outputs initialisation")) {
    return;
  }
  auto v = make_view(d);
  executePostProcessing(outputs, v, b, "PrincipalStrain");
  for (mgis::size_type i = 0; i != 3; ++i) {
    check(std::abs(outputs[i] - e2[i]) < eps, "invalid output value");
  }
}  // end of call_postprocessing

void call_postprocessing2(const mgis::behaviour::Behaviour& b) {
  using namespace mgis::behaviour;
  constexpr auto e =
      std::array<mgis::real, 6u>{1.3e-2, 1.2e-2, 1.4e-2, 0., 0., 0.};
  constexpr auto e2 =
      std::array<mgis::real, 6u>{1.2e-2, 1.3e-2, 1.4e-2, 0., 0., 0.};
  constexpr auto eps = 10 * std::numeric_limits<mgis::real>::epsilon();
  auto m = MaterialDataManager{b, 2u};
  // initialize the states
  setMaterialProperty(m.s1, "YoungModulus", 150e9);
  setMaterialProperty(m.s1, "PoissonRatio", 0.3);
  setExternalStateVariable(m.s1, "Temperature", 293.15);
  update(m);
  //
  for (mgis::size_type i = 0; i != 6; ++i) {
    m.s1.gradients[i] = e[i];
    m.s1.gradients[6 + i] = e[i];
  }
  //
  auto outputs = allocatePostProcessingVariables(m, "PrincipalStrain");
  if (!check(outputs.size() == 6u, "invalid outputs initialisation")) {
    return;
  }
  executePostProcessing(outputs, m, "PrincipalStrain");
  for (mgis::size_type i = 0; i != 3; ++i) {
    check(std::abs(outputs[i] - e2[i]) < eps, "invalid output value");
    check(std::abs(outputs[3 + i] - e2[i]) < eps, "invalid output value");
  }
}  // end of call_postprocessing2

int main(const int argc, const char* const* argv) {
  using namespace mgis::behaviour;
  constexpr const auto h = Hypothesis::TRIDIMENSIONAL;
  if (!check(argc == 3, "expected three arguments")) {
    return EXIT_FAILURE;
  }
  try {
    const auto b = load(argv[1], argv[2], h);
    check_behaviour(b, h);
    call_postprocessing(b);
    call_postprocessing2(b);
  } catch (std::exception& e) {
    std::cerr << e.what() << std::endl;
    return EXIT_FAILURE;
  }
  return success ? EXIT_SUCCESS : EXIT_FAILURE;
}  // end of main

File: bindings/python/tests/PostProcessingTest.py
# -*- coding: utf-8 -*-

import os
import numpy

try:
    import unittest2 as unittest
except ImportError:
    import unittest
import mgis.behaviour as mgis_bv


class PostProcessingTest(unittest.TestCase):
    def __get_behaviour(self, h):
        lib = os.environ['MGIS_TEST_BEHAVIOURS_LIBRARY']
        return mgis_bv.load(lib, 'PostProcessingTest', h)

    def test_behaviour(self):
        h = mgis_bv.Hypothesis.Tridimensional
        b = self.__get_behaviour(h)
        # path to the test library
        version = os.environ['MGIS_TEST_TFEL_VERSION']
        self.assertTrue(b.behaviour == "PostProcessingTest",
                        "invalid behaviour name")
        self.assertTrue(b.hypothesis == h, "invalid hypothesis")
        self.assertTrue(b.source == "PostProcessingTest.mfront",
                        "invalid source")
        self.assertTrue(b.tfel_version == version, "invalid TFEL version")
        postprocessings = b.getPostProcessingsNames()
        self.assertTrue(
            len(postprocessings) == 1, "invalid number of post-processings")
        self.assertTrue("PrincipalStrain" in postprocessings,
                        "invalid post-processing names")
        pass

    def test_behaviour_data(self):
        eps = 1.e-14
        e = numpy.asarray([1.3e-2, 1.2e-2, 1.4e-2, 0., 0., 0.],
                          dtype=numpy.float64)
        e2 = numpy.asarray([1.2e-2, 1.3e-2, 1.4e-2, 0., 0., 0.],
                           dtype=numpy.float64)
        h = mgis_bv.Hypothesis.Tridimensional
        b = self.__get_behaviour(h)
        m = mgis_bv.MaterialDataManager(b, 2)
        mgis_bv.setMaterialProperty(m.s1, "YoungModulus", 150e9)
        mgis_bv.setMaterialProperty(m.s1, "PoissonRatio", 0.3)
        mgis_bv.setExternalStateVariable(m.s1, "Temperature", 293.15)
        mgis_bv.update(m)
        m.s1.gradients[0:] = e
        m.s1.gradients[1:] = e
        print(m.s1.gradients)
        outputs = numpy.empty(shape=(2, 3), dtype=numpy.float64)
        mgis_bv.executePostProcessing(outputs.reshape(6), m, "PrincipalStrain")
        for i in range(0, 3):
            self.assertTrue(
                abs(outputs[0, i] - e2[i]) < eps,
                f"invalid output value ({outputs[0, i]} vs {e2[i]})")
            self.assertTrue(
                abs(outputs[1, i] - e2[i]) < eps,
                f"invalid output value ({outputs[1, i]} vs {e2[i]})")


if __name__ == '__main__':
    unittest.main()
